server/src/main/scala/org/http4s/netty/server/NegotiationHandler.scala

```diff
@@ -48,48 +37,24 @@ private[server] class NegotiationHandler[F[_]: Async](
   override def configurePipeline(ctx: ChannelHandlerContext, protocol: String): Unit =
     protocol match {
       case ApplicationProtocolNames.HTTP_2 =>
-        ctx
-          .pipeline()
-          .addLast(
-            Http2FrameCodecBuilder.forServer().build(),
-            new Http2MultiplexHandler(new ChannelInitializer[Channel] {
-              override def initChannel(ch: Channel): Unit = {
-                ch.pipeline().addLast(new Http2StreamFrameToHttpObjectCodec(true))
-                addToPipeline(ch.pipeline(), http1 = false)
-              }
-            })
-          )
-        ()
+        NettyPipelineHelpers.buildHttp2Pipeline(
+          ctx.pipeline,
+          config,
+          httpApp,
+          serviceErrorHandler,
+          dispatcher)
+
       case ApplicationProtocolNames.HTTP_1_1 =>
-        val pipeline = ctx.pipeline()
-        addToPipeline(pipeline, http1 = true)
-      case _ => throw new IllegalStateException(s"Protocol: $protocol not supported")
-    }
+        NettyPipelineHelpers.buildHttp1Pipeline(
+          ctx.pipeline,
+          config,
+          httpApp,
+          serviceErrorHandler,
+          dispatcher)
 
-  def addToPipeline(pipeline: ChannelPipeline, http1: Boolean): Unit = void {
-    if (http1) {
-      pipeline.addLast(
-        "http-decoder",
-        new HttpRequestDecoder(
-          config.maxInitialLineLength,
-          config.maxHeaderSize,
-          config.maxChunkSize))
-      pipeline.addLast("http-encoder", new HttpResponseEncoder())
+      case _ => throw new IllegalStateException(s"Protocol: $protocol not supported")
     }
 
-    if (config.idleTimeout.isFinite && config.idleTimeout.length > 0)
```

Comment:
Not fixed in this phab, but it looks like the idle timeout in HTTP/2 will only apply per-stream. I suspect we'd want it to apply for the whole session and also perhaps per-stream.